catalog/core/catalog-core-standardframework/src/test/java/ddf/catalog/cache/solr/impl/CachingFederationStrategyTest.java

```diff
@@ -0,0 +1,328 @@
+
+        WaitCondition.expect("Wait until the result appears in the response")
+                .until(() -> response.getHits(), is(HITS_RESULT))
+                .checkEvery(POLLING_INTERVAL, TimeUnit.MILLISECONDS)
+                .within(WITHIN_TIMEOUT, TimeUnit.SECONDS);
+    }
+
+    @Test
+    public void testFederateQueryAndUpdateCacheBlocking() throws Exception {
+        HashMap<String, Serializable> properties = new HashMap<>();
+        properties.put(QUERY_MODE, INDEX_QUERY_MODE);
+
+        QueryRequest fedQueryRequest = new QueryRequestImpl(mockQuery, properties);
+
+        MockSource mockSource = getMockSource(mockResponse);
+
+        List<Source> sources = Arrays.asList(mockSource);
+
+        QueryResponse response = strategy.federate(sources, fedQueryRequest);
+
+        WaitCondition.expect("Wait until the result appears in the response")
+                .until(() -> response.getHits(), is(HITS_RESULT))
+                .checkEvery(POLLING_INTERVAL, TimeUnit.MILLISECONDS)
+                .within(WITHIN_TIMEOUT, TimeUnit.SECONDS);
+    }
+
+    @Test
+    public void testFederateQueryAndUpdateCacheNotBlocking() throws Exception {
+        HashMap<String, Serializable> properties = new HashMap<>();
+        properties.put(QUERY_MODE, UPDATE_QUERY_MODE);
+
+        QueryRequest fedQueryRequest = new QueryRequestImpl(mockQuery, properties);
+
+        MockSource mockSource = getMockSource(mockResponse);
+
+        List<Source> sources = Arrays.asList(mockSource);
+
+        QueryResponse response = strategy.federate(sources, fedQueryRequest);
+
+        WaitCondition.expect("Wait until the result appears in the response")
+                .until(() -> response.getHits(), is(HITS_RESULT))
+                .checkEvery(POLLING_INTERVAL, TimeUnit.MILLISECONDS)
+                .within(WITHIN_TIMEOUT, TimeUnit.SECONDS);
+    }
+
+    @Test
+    public void testFederateQueryCacheWithOffsetHandler() throws Exception {
+
+        HashMap<String, Serializable> properties = new HashMap<>();
+        properties.put(QUERY_MODE, INDEX_QUERY_MODE);
+
+        // startIndex and pageSize must be > 1
+        Query mockQ = new QueryImpl(mock(NullFilterImpl.class),
+                2,
+                5,
+                mock(SortBy.class),
+                true,
+                1000);
+
+        QueryRequest fedQueryRequest = new QueryRequestImpl(mockQ, properties);
+
+        List<Source> sources = new ArrayList<>();
+
+        // Multiple sources needed for OffsetResultHandler to be used
+        for (int i = 0; i < 2; i++) {
+            sources.add(getMockSource(mockResponse));
+        }
+
+        QueryResponse response = strategy.federate(sources, fedQueryRequest);
+
+        // Response is changed by OffsetResultHandler, so getHits() returns 0
+        WaitCondition.expect("Wait until the result appears in the response")
+                .until(() -> response.getHits(), is(0L))
```

Comment:
Wouldn't the returning result still be 1? Might be better to have the two sources return different amount of results and assert on that. Source 1 has 1 result while source 2 has 2 results.